hypothesis-python/src/hypothesis/strategies/_internal/types.py

```diff
@@ -88,10 +88,44 @@ def type_sorting_key(t):
 
 
 def try_issubclass(thing, superclass):
-    thing = getattr(thing, "__origin__", None) or thing
-    superclass = getattr(superclass, "__origin__", None) or superclass
     try:
-        return issubclass(thing, superclass)
+        # Else we're looking at two distinct classes - which might be generics.
+        # That brings in some complications:
+        if issubclass(
+            getattr(thing, "__origin__", None) or thing,
+            getattr(superclass, "__origin__", None) or superclass,
+        ):
+            superclass_args = getattr(superclass, "__args__", None)
+            if sys.version_info[:2] == (3, 6) or not superclass_args:
+                # Python 3.6 doesn't have PEP-560 semantics, so stop early.
+                # The superclass is not generic, so we're definitely a subclass.
+                return True
+            # Sadly this is just some really fiddly logic to handle all the cases
+            # of user-defined generic types, types inheriting from parametrised
+            # generics, and so on.  If you need to change this code, read PEP-560
+            # and Hypothesis issue #2951 closely first, and good luck.  The tests
+            # will help you, I hope - good luck.
+            thing_args = getattr(thing, "__args__", None)
+            thing_orig_bases = getattr(thing, "__orig_bases__", None) or [None]
+            if thing_args is None and len(thing_orig_bases) != 1:  # pragma: no cover
+                raise ResolutionFailed(
+                    f"It looks like you're trying to resolve {thing!r} which "
+                    "inherits from more than one generic type, which isn't supported. "
+                    "Try using register_type_strategy()."
```

Comment:
Maybe point the user to create the issue (and add the URL in the error message)? I could imagine that bugs related to this logic might actually stem from using `register_type_strategy`.